### PR TITLE
Insertion of example in query/path/body

### PR DESCRIPTION
**Pull request by CristianoVotre**
Hi, I develop a form to put examples in the request, its was an old issue:
'Is there a way to provide example of request body? #180'

Now, we can put:
`processarDesconTurbo(
    @Path('idUsuario', 'ABCDEF125') idUsuario: string,
    @Path('idServico', '1') idServico: number,
    @Query('valorCompra', '19.90') valorCompra: number)
`
or in body tag:
`inserirPagamento(
    @Body(`{
        "usuario": {
          "codigo": "ABCDEF125"
        },
          "servico": {
          "codigo": 1
        },
        "desconTurbo": false,
        "parcela": 1,
        "valor": 19.90,
        "valorOnes": 19.90
      }`)
    pagamento: Pagamento)
`
And in the generated swagger will show these values on the fields.

### CODE REVIEW

**Comment by dgreene1** [comment]
Can you translate your text above? I apologize, but I can’t make sense if it.

File: package.json [hunk 1]
```diff
@@ -102,6 +103,9 @@
     "ts-node": "^8.3.0",
     "tslint": "^5.19.0"
   },
+  "publishConfig": {
```
Comment (dgreene1) [inline package.json]: What is this and why did you need to add it?

I’m not comfortable with using another registry other than npm.

File: src/decorators/parameter.ts [hunk 2]
```diff
@@ -2,7 +2,7 @@
  * Inject http Body
  *  @param {string} [name] properties name in body object
  */
-export function Body(): Function {
+export function Body(example?: any): Function {
```
Comment (dgreene1) [inline src/decorators/parameter.ts]: Don’t use `any` it should be at least a generic with a requirement to extend an object.

File: src/decorators/parameter.ts [hunk 3]
```diff
@@ -33,7 +33,7 @@ export function Request(): Function {
  *
  * @param {string} [name] The name of the path parameter
  */
-export function Path(name?: string): Function {
+export function Path(name?: string, example?: any): Function {
```
Comment (dgreene1) [inline src/decorators/parameter.ts]: Same as above. Don’t use any.

File: src/metadataGeneration/tsoa.ts [hunk 4]
```diff
@@ -30,12 +30,13 @@ export namespace Tsoa {
...
     parameterName: string;
     description?: string;
-    in: 'query' | 'header' | 'path' | 'formData' | 'body' | 'body-prop' | 'request';
+    in: 'query' | 'header' | 'path' | 'formData' | 'body' | 'body-prop' | 'request' | 'example';
     name: string;
     required?: boolean;
     type: Type;
     default?: any;
     validators: Validators;
+    example?: any;
```
Comment (dgreene1) [inline src/metadataGeneration/tsoa.ts]: Same as above. No any.

File: package.json (hunk 1 above)
Comment (CristianoVotre) [reply package.json]: Sorry, this file is wrong..., I will remove it.

File: src/metadataGeneration/tsoa.ts (hunk 4 above)
Comment (CristianoVotre) [reply src/metadataGeneration/tsoa.ts]: Ok

File: src/decorators/parameter.ts (hunk 3 above)
Comment (CristianoVotre) [reply src/decorators/parameter.ts]: Ok

File: src/decorators/parameter.ts (hunk 2 above)
Comment (CristianoVotre) [reply src/decorators/parameter.ts]: Ok

**Comment by dgreene1** [comment]
@CristianoVotre is this PR still necessary after [someone clarified that there is an alternative approach](https://github.com/lukeautry/tsoa/issues/180#issuecomment-526950724) that already exists?

See here: https://github.com/lukeautry/tsoa/issues/180#issuecomment-526950724

**Comment by WoH** [comment]
I believe this would still be a good addition since you can provide specific examples per path. (And would, therefore, be closer to the `@Request` decorator).
However, I'd like to note that OAPI3 does support multiple examples and I think, ideally, all the related decorators should accept multiple examples eventually (doesn't necessarily have to happen in this PR).

File: src/decorators/parameter.ts [hunk 5]
```diff
@@ -33,7 +33,7 @@ export function Request(): Function {
  *
  * @param {string} [name] The name of the path parameter
  */
-export function Path(name?: string): Function {
+export function Path(name?: string, example?: string): Function {
```
Comment (dgreene1) [inline src/decorators/parameter.ts]: `string` is not accurate either since path variables can be numbers, date, bool, etc (I believe)

File: src/decorators/parameter.ts [hunk 6]
```diff
@@ -2,7 +2,7 @@
  * Inject http Body
  *  @param {string} [name] properties name in body object
  */
-export function Body(): Function {
+export function Body(example?: string): Function {
```
Comment (dgreene1) [inline src/decorators/parameter.ts]: For all of these decorators, we need to make room for multiple configurations that could come down the line. So we need something extendable. That means an object. So instead of taking on primitive, instead have it be:

```ts

export interface BodyDecoratorConfig<T extends object> {
  swaggerOverride?: {
    description?: string,
    example?: T
  }
}

@Body(config?: BodyDecoratorConfig)
```

Comment (dgreene1) [reply src/decorators/parameter.ts]: What are your thoughts @WoH / @LukeAutry / @endor on the notion of allowing for people to override any of the Swagger properties? I ask this because this PR only allows the override of the example and I describe [here](https://github.com/lukeautry/tsoa/pull/459/files#r319777012) how we need to be able future proof by structuring the function to allow an object, for extensibility purposes: https://github.com/lukeautry/tsoa/pull/459/files#r319777012

Comment (WoH) [reply src/decorators/parameter.ts]: Good points. Easiest things first: Yes, every decorator should have a well-specified config interface. I am not against the idea of overriding specific high-level things on a per-request basis, but I'd have to take a look how far I'd go. For now I think, everything annotating the entire interface is reasonable (description and example). I've had cases where I'd have to duplicate an interface to adjust the example I'm giving (one that's valid in the context of the request).

Comment (WoH) [reply src/decorators/parameter.ts]: Now the harder thing: The OpenAPI spec says `example` is of type `any` which would be `unknown` in typescript.
However, it explicitly mentions: 

>  The example SHOULD match the specified schema and encoding properties if present. The example field is mutually exclusive of the examples field. Furthermore, if referencing a schema which contains an example, the example value SHALL override the example provided by the schema. To represent examples of media types that cannot naturally be represented in JSON or YAML, a string value can contain the example with escaping where necessary.

In order to get this checked, the config should allow a type argument so you are able to get type checking, with the fallback to `unknown` (Assuming we can't get typescript to infer the type of a decorated interface).

If we specify an example, we allow to specify all the properties of an example object though: https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#example-object

I'll have a look at this tomorrow, but these are the ideas at the top of my head.

Comment (CristianoVotre) [reply src/decorators/parameter.ts]: Well @WoH , I think its a good idea, I put the any value because that, but I can pass all the properties of object like in specification link and I can implement the structure like @dgreene1  suggested... I really need this...

Comment (WoH) [reply src/decorators/parameter.ts]: > I put the any value because that, but I can pass all the properties of object like in specification link

You can still accept everything if you ask for `unknown`, but TS will not allow you to do dangerous unchecked operations on the value.

However, before it makes sense for @dgreene1 to say yes here, it needs to generate correct Swagger 2 **and** OpenAPI3 specs. Also, Dan needs to think about the impact of this PR, regardless of the urgent requirement, he can't allow this PR to prevent future extensions of the features.

Comment (dgreene1) [reply src/decorators/parameter.ts]: > To represent examples of media types that cannot naturally be represented in JSON or YAML, a string value can contain the example with escaping where necessary.

This isn’t a problem for us because tsoa has an [established stance of only handling JSON](https://github.com/lukeautry/tsoa/issues/224#issuecomment-524644680). So I've decided that we should start with strict typing and we can relax it later if someone requests that we make a change/enhancement. Let’s leave the type of `example` as unknown. However, we can't allow tsoa users to make silly mistakes.

So we need to run `swaggerOverrides.example` through `ValidationService.ValidateParam()` during the generation to ensure the developer provided the correct type.

Why am I asking us to run validation instead of this idea of @WoH’s? (keep reading)

> In order to get this checked, the config should allow a type argument so you are able to get type checking, with the fallback to unknown (Assuming we can't get typescript to infer the type of a decorated interface).

So I would prefer that the TypeScript compiler inform the user that the example doesn’t match, but there is a [known issue with inference of parameter decorators](https://github.com/Microsoft/TypeScript/issues/30102) as illustrated by this SO question: https://stackoverflow.com/questions/55144265/how-to-limit-the-type-of-parameter-that-the-given-parameter-decorator-may-be-app).

Reminder @CristianoVotre, I still want the input to be a config object like this:
```ts
interface PathDecoratorConfig {
  name?: string,
  swaggerOverride: {
    example: unknown
  }
}

const parameterDecorator = (decoratorConfig?: string | PathDecoratorConfig) ...
```

Comment (WoH) [reply src/decorators/parameter.ts]: You can still check at compile time, but I think this would require a lot of work. If we'd want to have TS do the work, which is where I was looking at, I'd like to see something like `@Body<T>(config: Config<T>)` where `interface Config<T> = { description?: string, example: T }`. I disagree with the naming, since this is not overriding anything. These examples are explicit part of the OAPI3 spec.

Comment (dgreene1) [reply src/decorators/parameter.ts]: > You can still check at compile time, but I think this would require a lot of work. If we'd want to have TS do the work, which is where I was looking at, I'd like to see something like @Body<T>

I wish we could, but the problem is that `T` can’t be inferred for decorators of parameters. So that’s why [the StackOverflow issue](https://stackoverflow.com/questions/55144265/how-to-limit-the-type-of-parameter-that-the-given-parameter-decorator-may-be-app) presents this possible problem:

```ts
function postSomething( @Body<Thing1>() aParemeterWhosTypeIsNotInferred: Thing2 ): Thing2
```

^ that’s a really problem that the T of Body (I.e `Thing1`) can be different than the variable’s type (I.e. `Thing2`) that’s blocked by https://github.com/Microsoft/TypeScript/issues/30102

Comment (WoH) [reply src/decorators/parameter.ts]: > I wish we could, but the problem is that T can’t be inferred for decorators of parameters. 

It can, because we are in the fortunate situation to be able to parse the AST. However, since we both agreed we should not do it, there's no reason to discuss further imo.

> that’s a really unfortunate problem that the T of Body (I.e Thing1) can be different than the variable’s type (I.e. Thing2).

Yes. I'd say that's better than `unknown`, but I'm happy either way.

> No, swagger 2 supports example just not examples. 

This is incorrect in this context. This PR addresses query, path and body example(s). None of this is supported as is in Swagger 2 and therefore which the article you linked confirms by not mentioning any of these.

A[ Swagger 2] API specification can include examples for:
- response MIME types,
- schemas (data models),
- individual properties in schemas.

Comment (dgreene1) [reply src/decorators/parameter.ts]: Thank you for clarifying @WoH. What version of the spec are you using @CristianoVotre?
